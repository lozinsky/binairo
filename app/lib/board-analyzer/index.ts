export * from './analyze-board-by-kind';
export * from './analyze-board';
