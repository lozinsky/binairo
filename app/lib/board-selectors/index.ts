export * from './select-filled-imbalanced-line';
export * from './select-line-that-can-be-equal-to-another';
export * from './select-line-that-includes-another';
export * from './select-line-that-is-equal-to-others';
export * from './select-middle-cell-between-two-identical';
export * from './select-next-cell-after-two-identical';
export * from './select-three-or-more-identical-sequential-cells';
export * from './select-unfilled-balanced-line';
