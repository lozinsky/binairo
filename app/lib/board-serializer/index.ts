export * from './board-serializer-collection';
