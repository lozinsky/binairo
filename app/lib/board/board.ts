import {
  DoubleMatrixRotation,
  type Matrix,
  type MatrixReversible,
  type MatrixRotatable,
  type MatrixSelection,
  QuadrupleMatrixRotation,
} from '~/lib/matrix';
import { assert } from '~/shared/assert';
import { expectToBeDefined } from '~/shared/expect';

import type { BoardCell, BoardCellState } from './board-cell';

import { BoardLine, type BoardLineValue, isBoardLineValue } from './board-line';

export type BoardOrientation = 'landscape' | 'portrait';

const NEXT_BOARD_ORIENTATION_BY_PREV_BOARD_ORIENTATION: Readonly<Record<BoardOrientation, BoardOrientation>> = {
  landscape: 'portrait',
  portrait: 'landscape',
};

export type BoardValue = readonly BoardLineValue[];

export class Board implements Iterable<BoardLine>, Matrix<BoardLine>, MatrixReversible<Board>, MatrixRotatable<Board> {
  get length() {
    return this.#lines.length;
  }

  get orientation() {
    return this.#orientation;
  }

  get progress() {
    return this.#lines.reduce((progress, line) => progress + line.progress, 0) / this.#lines.length;
  }

  readonly #lines: readonly BoardLine[];
  readonly #orientation: BoardOrientation;

  constructor(orientation: BoardOrientation, lines: readonly BoardLine[]) {
    for (const line of lines) {
      assert(line.length === lines.length, 'Expected line length to be equal to board size');
    }

    this.#lines = lines;
    this.#orientation = orientation;
  }

  static blank(size: number) {
    return new this(
      'portrait',
      Array.from({ length: size }, () => BoardLine.blank(size)),
    );
  }

  static create(states: ReadonlyArray<readonly BoardCellState[]>) {
    return new this(
      'portrait',
      states.map((states) => BoardLine.create(states)),
    );
  }

  static from(value: BoardValue) {
    return new this(
      'portrait',
      value.map((value) => BoardLine.from(value)),
    );
  }

  at(index: number) {
    return this.#lines.at(index);
  }

  entries() {
    return this.#lines.entries();
  }

  replace(index: number, callback: (line: BoardLine) => BoardLine) {
    return new Board(this.#orientation, this.#lines.with(index, callback(expectToBeDefined(this.#lines.at(index)))));
  }

  replaceBy(selection: MatrixSelection, callback: (cell: BoardCell) => BoardCell) {
    const content = selection.execute(this);
    const lines = this.#lines.map((line) => {
      const cells = Array.from(line);

      for (const cell of content) {
        const index = line.indexOf(cell);

        if (index === -1) {
          continue;
        }

        cells[index] = callback(expectToBeDefined(cells[index]));
      }

      return new BoardLine(cells);
    });

    return new Board(this.#orientation, lines);
  }

  reverse() {
    return new Board(this.#orientation, this.#lines.toReversed());
  }

  rotate() {
    const orientation = NEXT_BOARD_ORIENTATION_BY_PREV_BOARD_ORIENTATION[this.#orientation];

    if (this.#lines.length === 0) {
      return new Board(orientation, []);
    }

    const lines: BoardLine[] = [];

    for (const index of expectToBeDefined(this.#lines[0]).keys()) {
      const cells: BoardCell[] = [];

      for (const line of this.#lines) {
        cells.push(expectToBeDefined(line.at(index)));
      }

      lines.push(new BoardLine(cells));
    }

    return new Board(orientation, lines);
  }

  slice(start: number, end?: number) {
    return this.#lines.slice(start, end);
  }

  [Symbol.iterator]() {
    return this.#lines[Symbol.iterator]();
  }

  toDoubleRotation() {
    return DoubleMatrixRotation.run<Board>(this);
  }

  toQuadrupleRotation() {
    return QuadrupleMatrixRotation.run<Board>(this);
  }

  valueOf(): BoardValue {
    return this.#lines.map((line) => line.valueOf());
  }
}

export function isBoardValue(value: unknown): value is BoardValue {
  return Array.isArray(value) && value.every(isBoardLineValue);
}
