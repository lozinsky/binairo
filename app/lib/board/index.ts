export * from './board-cell';
export * from './board-line';
export * from './board';
