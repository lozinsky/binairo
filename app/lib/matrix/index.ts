export * from './matrix-line';
export * from './matrix-rotation';
export * from './matrix-selection';
export * from './matrix';
