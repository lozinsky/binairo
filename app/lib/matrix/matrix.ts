import type { MatrixLine, MatrixLineCell } from './matrix-line';

export interface Matrix<L extends MatrixLine<unknown>> {
  at(index: number): L | undefined;
  entries(): Iterable<[number, L]>;
}

export type MatrixCell<M> = M extends Matrix<infer L> ? MatrixLineCell<L> : never;
