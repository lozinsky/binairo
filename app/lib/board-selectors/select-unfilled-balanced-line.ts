import { type Board, BoardCellState } from '~/lib/board';

export function selectUnfilledBalancedLine(target: Board) {
  for (const board of target.toDoubleRotation()) {
    for (const line of board) {
      if (!line.isFilled) {
        const length = Math.max(line.lengthOf(BoardCellState.R), line.lengthOf(BoardCellState.B));

        if (length === line.length / 2) {
          return { line, orientation: board.orientation };
        }
      }
    }
  }
}
