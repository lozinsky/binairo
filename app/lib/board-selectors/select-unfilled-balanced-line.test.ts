import { assert, expect, test } from 'vitest';

import { Board, BoardCellState } from '~/lib/board';
import { MatrixSelection } from '~/lib/matrix';

import { selectUnfilledBalancedLine } from './select-unfilled-balanced-line';

test.each([
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.R, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.R],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.R, BoardCellState.E, BoardCellState.E, BoardCellState.R],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.R, BoardCellState.R, BoardCellState.B, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.B, BoardCellState.R],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.B, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.B, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
  ]),
])('selects unfilled balanced line', (target) => {
  const payload = selectUnfilledBalancedLine(target);

  assert(payload !== undefined);

  const selection = MatrixSelection.from(target, Array.from(payload.line));

  expect(payload.orientation).toMatchSnapshot();
  expect(target.replaceBy(selection, (cell) => cell.toFixed())).toMatchSnapshot();
});

test.each([
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.R, BoardCellState.R, BoardCellState.R, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.R, BoardCellState.R, BoardCellState.E, BoardCellState.R],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.R, BoardCellState.E, BoardCellState.R, BoardCellState.R],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.R, BoardCellState.R, BoardCellState.B, BoardCellState.B],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.B, BoardCellState.R, BoardCellState.B, BoardCellState.R],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.B, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.B, BoardCellState.E, BoardCellState.E],
  ]),
  Board.create([
    [BoardCellState.E, BoardCellState.B, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.B, BoardCellState.E, BoardCellState.E],
    [BoardCellState.E, BoardCellState.R, BoardCellState.E, BoardCellState.E],
  ]),
])('returns "undefined" if no unfilled balanced line exist', (target) => {
  expect(selectUnfilledBalancedLine(target)).toMatchSnapshot();
});
