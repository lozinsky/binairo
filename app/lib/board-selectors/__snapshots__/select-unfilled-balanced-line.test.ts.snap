// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns "undefined" if no unfilled balanced line exist 1`] = `undefined`;

exports[`returns "undefined" if no unfilled balanced line exist 2`] = `undefined`;

exports[`returns "undefined" if no unfilled balanced line exist 3`] = `undefined`;

exports[`returns "undefined" if no unfilled balanced line exist 4`] = `undefined`;

exports[`returns "undefined" if no unfilled balanced line exist 5`] = `undefined`;

exports[`returns "undefined" if no unfilled balanced line exist 6`] = `undefined`;

exports[`returns "undefined" if no unfilled balanced line exist 7`] = `undefined`;

exports[`returns "undefined" if no unfilled balanced line exist 8`] = `undefined`;

exports[`returns "undefined" if no unfilled balanced line exist 9`] = `undefined`;

exports[`returns "undefined" if no unfilled balanced line exist 10`] = `undefined`;

exports[`selects unfilled balanced line 1`] = `1`;

exports[`selects unfilled balanced line 2`] = `
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
{ðŸŸ¥} {ðŸŸ¥} {â¬œ} {â¬œ}
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
`;

exports[`selects unfilled balanced line 3`] = `1`;

exports[`selects unfilled balanced line 4`] = `
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
{â¬œ} {ðŸŸ¥} {â¬œ} {ðŸŸ¥}
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
`;

exports[`selects unfilled balanced line 5`] = `1`;

exports[`selects unfilled balanced line 6`] = `
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
{ðŸŸ¥} {â¬œ} {â¬œ} {ðŸŸ¥}
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
`;

exports[`selects unfilled balanced line 7`] = `0`;

exports[`selects unfilled balanced line 8`] = `
[â¬œ] {ðŸŸ¥} [â¬œ] [â¬œ]
[â¬œ] {ðŸŸ¥} [â¬œ] [â¬œ]
[â¬œ] {â¬œ} [â¬œ] [â¬œ]
[â¬œ] {â¬œ} [â¬œ] [â¬œ]
`;

exports[`selects unfilled balanced line 9`] = `0`;

exports[`selects unfilled balanced line 10`] = `
[â¬œ] {â¬œ} [â¬œ] [â¬œ]
[â¬œ] {ðŸŸ¥} [â¬œ] [â¬œ]
[â¬œ] {â¬œ} [â¬œ] [â¬œ]
[â¬œ] {ðŸŸ¥} [â¬œ] [â¬œ]
`;

exports[`selects unfilled balanced line 11`] = `0`;

exports[`selects unfilled balanced line 12`] = `
[â¬œ] {ðŸŸ¥} [â¬œ] [â¬œ]
[â¬œ] {â¬œ} [â¬œ] [â¬œ]
[â¬œ] {â¬œ} [â¬œ] [â¬œ]
[â¬œ] {ðŸŸ¥} [â¬œ] [â¬œ]
`;

exports[`selects unfilled balanced line 13`] = `1`;

exports[`selects unfilled balanced line 14`] = `
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
{ðŸŸ¥} {ðŸŸ¥} {ðŸŸ¦} {â¬œ}
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
`;

exports[`selects unfilled balanced line 15`] = `1`;

exports[`selects unfilled balanced line 16`] = `
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
{â¬œ} {ðŸŸ¥} {ðŸŸ¦} {ðŸŸ¥}
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
[â¬œ] [â¬œ] [â¬œ] [â¬œ]
`;

exports[`selects unfilled balanced line 17`] = `0`;

exports[`selects unfilled balanced line 18`] = `
[â¬œ] {ðŸŸ¥} [â¬œ] [â¬œ]
[â¬œ] {ðŸŸ¥} [â¬œ] [â¬œ]
[â¬œ] {ðŸŸ¦} [â¬œ] [â¬œ]
[â¬œ] {â¬œ} [â¬œ] [â¬œ]
`;

exports[`selects unfilled balanced line 19`] = `0`;

exports[`selects unfilled balanced line 20`] = `
[â¬œ] {â¬œ} [â¬œ] [â¬œ]
[â¬œ] {ðŸŸ¥} [â¬œ] [â¬œ]
[â¬œ] {ðŸŸ¦} [â¬œ] [â¬œ]
[â¬œ] {ðŸŸ¥} [â¬œ] [â¬œ]
`;
